### PR TITLE
ERR: Improve error message on non-sorted input with .truncate

### LINKED ISSUE #17935: ERR: better error message on non-sorted input with .truncate
xref #17295

This is correct, but since the doc-string says sorted, we should check this and give a nicer message. (not that should check both ``.is_monotonic_increasing`` and ``decreasing`` cases, and confirm that the before/after match)
```
In [8]: rng2 = pd.date_range('2011-01-01', '2012-01-01', freq='W')
   ...: ts2 = pd.Series(np.random.randn(len(rng2)), index=rng2)

In [9]: ts2.truncate(before='2011-11', after='2011-12')
Out[9]: 
2011-11-06    0.144986
2011-11-13   -1.823425
2011-11-20    1.742315
2011-11-27    0.311337
Freq: W-SUN, dtype: float64

In [10]: ts2.sort_values(ascending=False).truncate(before='2011-11', after='2011-12')
KeyError: Timestamp('2011-11-01 00:00:00')
```

### PR DESCRIPTION
**Pull request by reidy-p**
- [ ] closes #17935
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by gfyoung — APPROVED** [review]

**Review by TomAugspurger — APPROVED** [review]

File: pandas/tests/frame/test_timeseries.py [hunk 1]
```diff
@@ -374,6 +374,31 @@ def test_truncate_copy(self):
         truncated.values[:] = 5.
         assert not (self.tsframe.values[5:11] == 5).any()
 
+    def test_truncate_nonsortedindex(self):
```
Comment (jreback) [inline pandas/tests/frame/test_timeseries.py]: can you add a test for series as well (you can also put this test generically in test_generic if its easier)

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a note in other api changes in 0.22 (this is technially an api change, sort of a trivial one)

File: pandas/core/generic.py [hunk 2]
```diff
@@ -6397,6 +6397,10 @@ def truncate(self, before=None, after=None, axis=None, copy=True):
         axis = self._get_axis_number(axis)
         ax = self._get_axis(axis)
```
Comment (jreback) [inline pandas/core/generic.py]: add a comment here

**Comment by jreback** [comment]
can you rebase

File: pandas/core/generic.py [hunk 3]
```diff
@@ -6338,6 +6338,12 @@ def truncate(self, before=None, after=None, axis=None, copy=True):
...
 
+        # GH 17935
+        # Check that index is sorted
+        if (not ax.is_monotonic_increasing and
+           not ax.is_monotonic_decreasing):
```
Comment (jreback) [inline pandas/core/generic.py]: lint error

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, except you have lint issues, you can run ``make lint-diff`` and see

File: doc/source/whatsnew/v0.22.0.txt [hunk 4]
```diff
@@ -46,6 +46,8 @@ Other API Changes
 - :class:`Timestamp` will no longer silently ignore invalid ``freq`` arguments (:issue:`5168`)
 - :class:`CacheableOffset` and :class:`WeekDay` are no longer available in the ``pandas.tseries.offsets`` module (:issue:`17830`)
 - `tseries.frequencies.get_freq_group()` and `tseries.frequencies.DAYS` are removed from the public API (:issue:`18034`)
+- :func:`Series.truncate` and :func:`DataFrame.truncate` will raise a ``ValueError`` if the index is not sorted (:issue:`17935`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: say, rather than a non-helpful ``KeyError``

**Comment by reidy-p** [comment]
I updated the whatsnew and think I fixed the lint issue.

But since I opened this PR a new the ``test_truncate`` test has been added to ``pandas/tests/series/test_period.py`` in #17755. This new test fails on my branch because it tries to truncate on a non-sorted index. So I modified the test to sort the index before truncating and now it passes. Alternatively, I could simply check for the new error message ``ValueError: truncate requires a sorted index`` in this test.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @reidy-p 

changing the test for Period was just fine

**Pull request merged**